### ISSUE TITLE
Ability to analyze AST before applying rules

### ISSUE THREAD

**Issue by jacksonrayhamilton** [eslint/eslint#4556]
&#49;. The version of ESLint I am using: 3d712c3
&#50;. The problem I want to solve: I'd like to be able to write an extension to ESLint that analyzes the AST and sets configuration before applying rules. This is because I'd like to integrate ESLint with a preprocessor, leveraging that preprocessor's comment syntax to create inline linting configuration. For instance:

```js
// import a from ./foo/bar.js
a.c();
```

In the above snippet, I'd like to configure ESLint to search the comments in my code for `// import` and set `a` as a global / local variable in that script.

It'd also be interesting if I could configure ESLint to look at `./foo/bar.js` to see if there is a `var a` statement somewhere in it, and throw a linter error if there isn't.

&#51;. My take on the correct solution to problem: I don't know. Is this already possible? The plugin API seems to be designed for creating rules. This doesn't seem like a good fit as a rule, as this wouldn't really check anything, it would just modify the config as a side-effect of running. It would also need to run before other rules, which might violate the "every rule is standalone" assertion on the website's "About" page.

**Comment by eslintbot** [comment]
@jacksonrayhamilton Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by jacksonrayhamilton** [comment]
Sorry, I no longer have a need for this issue to be resolved.

**Comment by jacksonrayhamilton** [comment]
Interest revived!

**Comment by nzakas** [comment]
You might be able to use processors to accomplish something like that: http://eslint.org/docs/developer-guide/working-with-plugins.html#processors-in-plugins

It only works on text, though, so your best bet would be to replace the comment with a var statement.

Otherwise, what you're asking is not possible and we are unlikely to add such an enhancement anytime soon.

**Comment by ilyavolodin** [comment]
@jacksonrayhamilton You can already do this as is. You would need to write a processor, that can extract `// import` statements through RegEx, and substitute it with either `var a = ...` or `//globals a`.
